### PR TITLE
Kp/gcn experiments

### PR DESCRIPTION
**Pull request by kprokofi**

### CODE REVIEW

File: scripts/default_config.py [hunk 1]
```diff
@@ -28,7 +28,7 @@ def get_default_config():
     cfg.lr_finder.min_lr = 0.004
     cfg.lr_finder.step = None
     cfg.lr_finder.num_epochs = 3
-    cfg.lr_finder.epochs_warmup = 2
```
Comment (sovrasov) [inline scripts/default_config.py:31]: Will this change affect non-gcn training scenarios?

File: torchreid/data/datasets/dataset.py [hunk 2]
```diff
@@ -67,6 +67,7 @@ def __getitem__(self, index):
     def __len__(self):
         return len(self.data)
 
+
```
Comment (sovrasov) [inline torchreid/data/datasets/dataset.py]: Useless new line

File: torchreid/engine/engine.py [hunk 3]
```diff
@@ -558,6 +571,73 @@ def train(self, print_freq=10, fixbase_epoch=0, open_layers=None, lr_finder=Fals
...
+            bby1 = np.clip(cy - cut_h // 2, 0, H)
+            bbx2 = np.clip(cx + cut_w // 2, 0, W)
+            bby2 = np.clip(cy + cut_h // 2, 0, H)
+
+            return bbx1, bby1, bbx2, bby2
+
+        if self.aug_type == 'fmix':
+            r = np.random.rand(1)
+            if self.alpha > 0 and r[0] <= self.aug_prob:
+                lam, fmask = sample_mask(self.alpha, self.decay_power, self.size)
```
Comment (sovrasov) [inline torchreid/engine/engine.py]: Size could be inferred from images

File: torchreid/engine/image/multilabel.py [hunk 4]
```diff
@@ -17,7 +18,7 @@ class MultilabelEngine(Engine):
     r"""Multilabel classification engine. It supports ASL, BCE and Angular margin loss with binary classification."""
     def __init__(self, datamanager, models, optimizers, schedulers, use_gpu, save_all_chkpts,
                  train_patience, early_stopping, lr_decay_factor, loss_name, label_smooth,
-                 lr_finder, m, amb_k, amb_t, clip_grad,
+                 lr_finder, m, amb_k, amb_t, clip_grad, aug_prob, alpha, aug_type,
```
Comment (sovrasov) [inline torchreid/engine/image/multilabel.py:21]: The new params are not used in the constructor

File: torchreid/losses/asl.py [hunk 5]
```diff
@@ -10,7 +10,7 @@ def __init__(self, gamma_neg=4, gamma_pos=0,
                     probability_margin=0.05, eps=1e-8,
                     label_smooth=0.):
         super().__init__()
-
+        self.bce = nn.MultiLabelSoftMarginLoss()
```
Comment (sovrasov) [inline torchreid/losses/asl.py]: Useless variable

File: torchreid/models/__init__.py [hunk 6]
```diff
@@ -55,8 +57,20 @@ def build_model(name, **kwargs):
...
+    elif name.startswith('gcn'):
+        backbone_name = name[4:]
+        if backbone_name not in avai_models:
+            raise KeyError('Unknown backbone for GCN model: {}. Must be one of {}'.format(backbone_name, avai_models))
+        backbone = __model_factory[backbone_name](**kwargs)
+        model = build_image_gcn(backbone, **kwargs)
+    elif name.startswith('mld'):
+        backbone_name = name[4:]
+        if backbone_name not in avai_models:
+            raise KeyError('Unknown backbone for Q2L model: {}. Must be one of {}'.format(backbone_name, avai_models))
```
Comment (sovrasov) [inline torchreid/models/__init__.py]: ... for MLD

File: torchreid/models/common.py [hunk 7]
```diff
@@ -60,6 +60,19 @@ def get_num_features(self):
     def get_num_head_features(self):
         return self.num_head_features
 
+    def get_config_optim(self, lrs):
```
Comment (sovrasov) [inline torchreid/models/common.py]: Why do we need two copies of this method here?

File: torchreid/models/common.py [hunk 8]
```diff
@@ -69,7 +82,7 @@ def compute_s(num_class: int):
         return float(min(max(np.sqrt(5) * np.log(num_class - 1), 7), 30.))
 
     @staticmethod
-    def _glob_feature_vector(x, mode, reduce_dims=True):
+    def glob_feature_vector(x, mode, reduce_dims=True):
```
Comment (sovrasov) [inline torchreid/models/common.py]: Why did you remove underscore?

File: torchreid/models/ml_decoder.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
+from typing import Optional
```
Comment (sovrasov) [inline torchreid/models/ml_decoder.py:4]: I'd place the original license here  as Leonid did for other files. The same is for GCN.

File: torchreid/losses/asl.py [hunk 10]
```diff
@@ -77,6 +76,17 @@ def __init__(self, m=0.35, k=0.8, t=1, s=30,
         # prevent memory allocation and gpu uploading every iteration, and encourages inplace operations
         self.anti_targets = self.xs_pos = self.xs_neg = self.asymmetric_w = self.loss = None
 
+    @staticmethod
+    def _valid(params):
+        if isinstance(params, (list, tuple)):
```
Comment (sovrasov) [inline torchreid/losses/asl.py]: Is aug_index a hierarchical structure?

**Comment by sovrasov** [comment]
Task and configs were removed from this repo, please move the changes to ote: https://github.com/openvinotoolkit/training_extensions/pull/972

**Comment by kprokofi** [comment]
> Task and configs were removed from this repo, please move the changes to ote: [openvinotoolkit/training_extensions#972](https://github.com/openvinotoolkit/training_extensions/pull/972)

Do you mean just merge or rebase to the current state of the ote branch?

File: torchreid/models/common.py (hunk 7 above)
Comment (kprokofi) [reply torchreid/models/common.py]: I can see only one method here. What do you mean by "two copies"?

**Comment by sovrasov** [comment]
You could try to revert changes in configs. This would eliminate the merge conflict

File: torchreid/models/common.py (hunk 7 above)
Comment (kprokofi) [reply torchreid/models/common.py]: Sorry, found it. I will remove this.

File: torchreid/losses/asl.py (hunk 10 above)
Comment (kprokofi) [reply torchreid/losses/asl.py]: No, but we pass `logits_aug_avai = self._valid([aug_index, lam])`, a list of parameters and we iterate over this list to check each parameter

File: scripts/default_config.py (hunk 1 above)
Comment (kprokofi) [reply scripts/default_config.py:31]: It is just for fast-ai pretraining. 1 epoch is enough to stabilize weights before putting the model to the algorithm.

File: torchreid/models/ml_decoder.py (hunk 9 above)
Comment (kprokofi) [reply torchreid/models/ml_decoder.py:4]: for GCN there is no license at all

Comment (kprokofi) [reply torchreid/models/ml_decoder.py:4]: I added MIT

**Comment by kprokofi** [comment]
I haven't managed with rebase. It is too dirty with commits
please, find new clean PR: https://github.com/openvinotoolkit/deep-object-reid/pull/209
